File: backend/app/services/cache_service.py
# app/services/cache_service.py

import asyncio
import functools
import hashlib
import json
import uuid
from collections.abc import Callable
from typing import Any

from app.core.config import SETTING
from app.core.database import get_redis
from app.middleware.logging import logger
from app.utils.codec import CacheCodec
from app.utils.renew_lock import RenewLock


def generate_cache_key(
    args: list[Any] | None = None,
    kwargs: dict[str, Any] | None = None,
    exclude_args: list[int] | None = None,
    exclude_kwargs: list[str] | None = None,
    key_prefix: str | None = None,
) -> str:
    """
    生成缓存 key

    :param args: 用于生成 key 的位置参数
    :param kwargs: 用于生成 key 的关键字参数
    :param exclude_args: 不参与 key 生成的位置参数索引
    :param exclude_kwargs: 不参与 key 生成的关键字参数名称
    :param key_prefix: key 前缀
    :return: 缓存 key
    """
    args = args or []
    kwargs = kwargs or {}
    exclude_args = exclude_args or []
    exclude_kwargs = exclude_kwargs or []

    filtered_args = [arg for i, arg in enumerate(args) if i not in exclude_args]
    filtered_kwargs = {k: v for k, v in kwargs.items() if k not in exclude_kwargs}

    serialized = json.dumps(
        {"args": filtered_args, "kwargs": filtered_kwargs},
        sort_keys=True,
        default=str,
    )

    key_hash = hashlib.md5(serialized.encode()).hexdigest()
    return f"{key_prefix}:{key_hash}"


async def cache_get(
    *,
    args: list[Any] | None = None,
    kwargs: dict[str, Any] | None = None,
    expire: int = 300,
    ignore_null: bool = True,
    exclude_args: list[int] | None = None,
    exclude_kwargs: list[str] | None = None,
    key_prefix: str | None = None,
    lock_timeout: int = SETTING.FALL_BACK_FUNCTION_WINDOW,
    fallback_func: Callable[..., Any] | None = None,
    codec: CacheCodec = CacheCodec(),
) -> Any:
    """
    获取缓存

    :param args: 用于生成缓存 key 的位置参数
    :param kwargs: 用于生成缓存 key 的关键字参数
    :param expire: 缓存过期时间(秒)
    :param ignore_null: 如果返回值为 None 或空列表,则不缓存,默认值True表示不会存入空值
    :param exclude_args: 不参与缓存 key 生成的位置参数索引
    :param exclude_kwargs: 不参与缓存 key 生成的关键字参数名称
    :param key_prefix: 缓存 key 的前缀
    :param lock_timeout: 锁的过期时间(秒)
    :param fallback_func: 缓存获取失败执行的回源函数
    :param codec: 缓存值编解码器
    """
    cache_key = generate_cache_key(
        args, kwargs, exclude_args, exclude_kwargs, key_prefix
    )
    lock_key = f"lock:{cache_key}"
    redis_pool = await get_redis()

    # 1. 尝试读缓存
    try:
        cached = await redis_pool.get(cache_key)
        if cached is not None:
            logger.info(f"Cache hit: {cache_key}")
            return codec.decode(cached)
    except Exception as e:
        logger.error(f"Cache read failed: {e}")

    # 2. 如果没有 fallback_func,直接返回 None
    if fallback_func is None:
        return None

    # 3. 有 fallback_func:尝试加锁回源
    lock_acquired = False
    lock_value = str(uuid.uuid4())
    renew_lock = None

    try:
        # 尝试获取锁(带唯一值)
        lock_acquired = await redis_pool.set(
            lock_key, lock_value, ex=lock_timeout, nx=True
        )

        if lock_acquired:
            # 启动锁续期任务
            renew_lock = RenewLock(lock_key, lock_value, lock_timeout)
            renew_lock.start()
            # 双重检查:可能在加锁前已被写入
            cached = await redis_pool.get(cache_key)
            if cached is not None:
                return codec.decode(cached)

            # 执行回源逻辑
            result = await fallback_func()

            # 决定是否缓存
            should_cache = not (
                ignore_null and (result is None or result == {} or result == [])
            )
            if should_cache:
                await redis_pool.setex(cache_key, expire, codec.encode(result))
                logger.info(f"Cache set: {cache_key} (expire={expire}s)")

            return result
        else:
            # 未获取到锁:循环等待缓存被写入
            start_time = asyncio.get_event_loop().time()
            max_wait = lock_timeout + 1  # 略大于锁超时,防误判

            while asyncio.get_event_loop().time() - start_time < max_wait:
                await asyncio.sleep(0.05)
                cached = await redis_pool.get(cache_key)
                if cached is not None:
                    logger.info(f"Cache filled by another worker: {cache_key}")
                    return codec.decode(cached)

            # 超时仍未命中,认为锁持有者失败,自己回源
            logger.warning(
                f"Cache still empty after {max_wait}s, "
                f"executing fallback directly: {cache_key}"
            )
            return await fallback_func()

    except Exception as e:
        logger.error(f"Error in cache_get fallback: {e}")
        raise

    finally:
        if renew_lock is not None:
            await renew_lock.stop()

        # 安全释放锁(仅删除属于自己的锁)
        if lock_acquired:
            try:
                lua_script = """
                if redis.call("get", KEYS[1]) == ARGV[1] then
                    return redis.call("del", KEYS[1])
                else
                    return 0
                end
                """
                await redis_pool.eval(lua_script, 1, lock_key, lock_value)
            except Exception as e:
                logger.warning(f"Failed to release lock {lock_key}: {e}")


async def cache_set(
    *,
    value: Any,
    args: list[Any] | None = None,
    kwargs: dict[str, Any] | None = None,
    expire: int = 300,
    ignore_null: bool = True,
    exclude_args: list[int] | None = None,
    exclude_kwargs: list[str] | None = None,
    key_prefix: str | None = None,
    lock_timeout: int = SETTING.FALL_BACK_FUNCTION_WINDOW,
    acquire_lock: bool = True,  # 新增:是否尝试加锁
    codec: CacheCodec = CacheCodec(),
) -> bool:
    """
    设置缓存

    :param value: 缓存的值
    :param args: 用于生成缓存 key 的位置参数
    :param kwargs: 用于生成缓存 key 的关键字参数
    :param expire: 缓存过期时间(秒)
    :param ignore_null: 如果值为 None 或空列表,则不缓存,默认值True表示不会存入空值
    :param exclude_args: 不参与缓存 key 生成的位置参数索引
    :param exclude_kwargs: 不参与缓存 key 生成的关键字参数名称
    :param key_prefix: 缓存 key 的前缀
    :param lock_timeout:函数执行过期时间(秒)
    :param acquire_lock: 是否加锁,如果True,则在lock_timeout时间内只有一个线程能够访问
    :param codec: 缓存值编解码器
    """
    if key_prefix is None or key_prefix == "":
        raise ValueError("key_prefix must be set")
    if ignore_null and (value is None or value == {} or value == []):
        return False

    cache_key = generate_cache_key(
        args, kwargs, exclude_args, exclude_kwargs, key_prefix
    )
    redis_pool = await get_redis()
    # 如果不需要锁,直接设置
    if not acquire_lock:
        try:
            await redis_pool.setex(cache_key, expire, codec.encode(value))
            logger.info(f"Manual cache set (no lock): {cache_key} (expire={expire}s)")
            return True
        except Exception as e:
            logger.error(f"Manual cache set failed: {e}")
            return False

    # === 以下为带锁逻辑(与 cache_get 兼容)===
    lock_key = f"lock:{cache_key}"
    lock_value = str(uuid.uuid4())
    lock_acquired = False

    try:
        # 尝试获取锁
        lock_acquired = await redis_pool.set(
            lock_key, lock_value, ex=lock_timeout, nx=True
        )
        if not lock_acquired:
            logger.debug(
                f"cache_set skipped: lock held by another process for {cache_key}"
            )
            return False

        # 双重检查:可能已有新值(比如 cache_get 刚写入)
        existing = await redis_pool.get(cache_key)
        if existing is not None:
            logger.debug(f"cache_set skipped: cache already exists for {cache_key}")
            return False

        # 设置缓存
        await redis_pool.setex(cache_key, expire, codec.encode(value))
        logger.info(f"Manual cache set with lock: {cache_key} (expire={expire}s)")
        return True

    except Exception as e:
        logger.error(f"Manual cache set with lock failed: {e}")
        return False

    finally:
        # 安全释放锁(仅删除属于自己的)
        if lock_acquired:
            try:
                lua_script = """
                if redis.call("get", KEYS[1]) == ARGV[1] then
                    return redis.call("del", KEYS[1])
                else
                    return 0
                end
                """
                await redis_pool.eval(lua_script, 1, lock_key, lock_value)
            except Exception as e:
                logger.warning(f"Failed to release lock in cache_set {lock_key}: {e}")


async def cache_delete(
    *,
    args: list[Any] | None = None,
    kwargs: dict[str, Any] | None = None,
    exclude_args: list[int] | None = None,
    exclude_kwargs: list[str] | None = None,
    key_prefix: str | None = None,
) -> bool:
    """
    删除缓存

    :param args: 用于生成缓存 key 的位置参数
    :param kwargs: 用于生成缓存 key 的关键字参数
    :param exclude_args: 不参与缓存 key 生成的位置参数索引
    :param exclude_kwargs: 不参与缓存 key 生成的关键字参数名称
    :param key_prefix: 缓存 key 的前缀
    """
    if key_prefix is None or key_prefix == "":
        raise ValueError("key_prefix must be set")
    redis_pool = await get_redis()
    try:
        cache_key = generate_cache_key(
            args, kwargs, exclude_args, exclude_kwargs, key_prefix
        )
        result = await redis_pool.delete(cache_key)
        deleted = result > 0
        if deleted:
            logger.info(f"Cache deleted: {cache_key}")
        else:
            logger.info(f"Cache key not found for deletion: {cache_key}")
        return deleted
    except Exception as e:
        logger.error(f"Cache delete failed: {e}")
        return False


def cache(
    *,
    expire: int = 300,
    ignore_null: bool = True,
    exclude_args: list[int] | None = None,
    exclude_kwargs: list[str] | None = None,
    key_prefix: str | None = None,
    lock_timeout: int = SETTING.FALL_BACK_FUNCTION_WINDOW,
    codec: CacheCodec = CacheCodec(),
):
    """
    缓存装饰器

    :param expire: 缓存过期时间(秒)
    :param ignore_null: 如果值为 None 或空列表,则不缓存,默认值True表示不会存入空值
    :param exclude_args: 不参与缓存 key 生成的位置参数索引
    :param exclude_kwargs: 不参与缓存 key 生成的关键字参数名称
    :param key_prefix: 缓存 key 的前缀
    :param lock_timeout:函数执行过期时间(秒)
    :param codec: 缓存值编解码器,默认不做特殊处理
    """

    def decorator(func: Callable) -> Callable:
        @functools.wraps(func)
        async def wrapper(*args, **kwargs) -> Any:
            async def _fallback():
                return await func(*args, **kwargs)

            return await cache_get(
                args=list(args),
                kwargs=kwargs,
                expire=expire,
                ignore_null=ignore_null,
                exclude_args=exclude_args,
                exclude_kwargs=exclude_kwargs,
                key_prefix=key_prefix if key_prefix else func.__qualname__,
                lock_timeout=lock_timeout,
                fallback_func=_fallback,
                codec=codec,
            )

        return wrapper

    return decorator

File: ktorBackend/src/main/kotlin/com/qianrenni/services/CacheService.kt
package com.qianrenni.services

import com.qianrenni.database.redisManager
import io.ktor.server.application.*
import io.ktor.server.routing.*
import io.ktor.util.*
import io.lettuce.core.ScriptOutputType
import io.lettuce.core.api.async.RedisAsyncCommands
import kotlinx.coroutines.*
import kotlinx.coroutines.future.await
import kotlinx.serialization.KSerializer
import kotlinx.serialization.json.Json
import java.security.MessageDigest
import java.util.*
import kotlin.time.Duration.Companion.milliseconds
import kotlin.time.Duration.Companion.seconds

class CacheService(val application: Application) {

    private val json = Json { ignoreUnknownKeys = true; encodeDefaults = true }

    private fun md5(input: String): String {
        val md = MessageDigest.getInstance("MD5")
        return md.digest(input.toByteArray()).joinToString("") { "%02x".format(it) }
    }

    fun generateCacheKey(
        args: List<Any> = emptyList(),
        excludeArgs: List<Int> = emptyList(),
        keyPrefix: String
    ): String {
        val filteredArgs = args.filterIndexed { index, _ -> index !in excludeArgs }
        // 将参数转为字符串再序列化，避免 kotlinx.serialization 对 Any? 类型的限制
        val payload = filteredArgs.map { it.toString() }.joinToString("")

        return "$keyPrefix:${md5(payload)}"
    }

    suspend fun <T> cacheGetInternal(
        cacheKey: String,
        expire: Int,
        ignoreNull: Boolean,
        lockTimeout: Int,
        fallbackFunc: suspend () -> T,
        serializer: KSerializer<T>,
        redis: RedisAsyncCommands<String, String>
    ): T {
        // 1. 尝试读缓存
        val cached = redis.get(cacheKey).await()
        if (cached != null) {
            application.log.info("Cache hit: $cacheKey")
            return json.decodeFromString(serializer, cached)
        }
        // 2. 尝试加锁回源
        val lockKey = "lock:$cacheKey"
        val lockValue = UUID.randomUUID().toString()
        var lockAcquired = false
        var renewJob: Job? = null

        try {
            // Lettuce 设置 NX 和 EX
            val setResult = redis.setex(lockKey, lockTimeout.toLong(), lockValue).await()
            lockAcquired = setResult != null

            if (lockAcquired) {
                // 启动协程进行锁续期 (Watchdog)
                renewJob = CoroutineScope(SupervisorJob() + Dispatchers.Default).launch {
                    while (isActive) {
                        delay(lockTimeout.seconds) // 每 1/3 超时时间续期
                        try {
                            redis.expire(lockKey, lockTimeout.toLong()).await()
                        } catch (e: Exception) {
                            application.log.error("Lock renew failed: $e")
                            break
                        }
                    }
                }

                // 双重检查
                val cached = redis.get(cacheKey).await()
                if (cached != null) return json.decodeFromString(serializer, cached)

                // 执行回源逻辑
                val result = fallbackFunc()

                // 决定是否缓存
                val isEmpty = result == null ||
                        (result is Collection<*> && result.isEmpty()) ||
                        (result is Map<*, *> && result.isEmpty())

                if (!(ignoreNull && isEmpty)) {
                    val encoded = json.encodeToString(serializer, result)
                    redis.setex(cacheKey, expire.toLong(), encoded).await()
                    application.log.info("Cache set: $cacheKey (expire=${expire}s)")
                }
                return result
            } else {
                // 未获取到锁：循环等待
                val startTime = System.currentTimeMillis()
                val maxWait = (lockTimeout + 1) * 1000L

                while (System.currentTimeMillis() - startTime < maxWait) {
                    delay(50.milliseconds)
                    val cached = redis.get(cacheKey).await()
                    if (cached != null) {
                        application.log.info("Cache filled by another worker: $cacheKey")
                        return json.decodeFromString(serializer, cached)
                    }
                }

                // 超时直接回源
                application.log.warn("Cache still empty after ${maxWait}ms, executing fallback directly: $cacheKey")
                return fallbackFunc()
            }
        } catch (e: Exception) {
            application.log.error("Error in cacheGet fallback: $e")
            throw e
        } finally {
            renewJob?.cancel() // 停止续期
            if (lockAcquired) {
                try {
                    // Lua 脚本安全释放锁
                    val luaScript = """
                        if redis.call("get", KEYS[1]) == ARGV[1] then
                            return redis.call("del", KEYS[1])
                        else
                            return 0
                        end
                    """
                    redis.eval<Long>(luaScript, ScriptOutputType.INTEGER, arrayOf(lockKey), lockValue).await()
                } catch (e: Exception) {
                    application.log.warn("Failed to release lock $lockKey: $e")
                }
            }
        }
    }

    /**
     * 简单设置缓存（不带锁）
     */
    suspend fun cacheSetSimple(
        key: String,
        value: String,
        expire: Long,
    ) {
        val redis = application.redisManager.getAsyncCommands()
        try {
            redis.setex(key, expire, value).await()
        } catch (e: Exception) {
            application.log.error("Error in cache set: $key: $e")
            throw IllegalStateException("服务器繁忙，请稍后再试")
        }
    }

    /**
     * 简单获取缓存值（字符串）
     */
    suspend fun cacheGetSimple(key: String): String? {
        val redis = application.redisManager.getAsyncCommands()
        return try {
            redis.get(key).await()
        } catch (e: Exception) {
            application.log.error("Cache get failed: $e")
            null
        }
    }

    /**
     * 简单删除缓存
     */
    suspend fun cacheDelete(key: String): Boolean {
        val redis = application.redisManager.getAsyncCommands()
        return try {
            redis.del(key).await() > 0
        } catch (e: Exception) {
            application.log.error("Cache delete failed: $e")
            false
        }
    }

    /**
     * 手动设置缓存
     */
    suspend fun <T> cacheSet(
        value: T,
        args: List<Any> = emptyList(),
        expire: Int = 300,
        ignoreNull: Boolean = true,
        excludeArgs: List<Int> = emptyList(),
        keyPrefix: String,
        lockTimeout: Int = 30,
        acquireLock: Boolean = true,
        serializer: KSerializer<T>,
    ): Boolean {
        val isEmpty = (value is Collection<*> && value.isEmpty()) || (value is Map<*, *> && value.isEmpty())
        if (ignoreNull && isEmpty) return false

        val redis = application.redisManager.getAsyncCommands()
        val cacheKey = generateCacheKey(args, excludeArgs, keyPrefix)
        val encoded = json.encodeToString(serializer, value)
        if (!acquireLock) {
            return try {
                redis.setex(cacheKey, expire.toLong(), encoded).await()
                true
            } catch (e: Exception) {
                application.log.error("Manual cache set failed: $e")
                false
            }
        }

        // 带锁逻辑 (与 cacheGet 类似，此处简化展示核心逻辑)
        val lockKey = "lock:$cacheKey"
        val lockValue = UUID.randomUUID().toString()
        var lockAcquired = false
        try {
            val setResult = redis.setex(lockKey, lockTimeout.toLong(), lockValue).await()
            lockAcquired = setResult != null
            if (!lockAcquired) return false
            // 双重检查
            if (redis.get(cacheKey).await() != null) return false
            redis.setex(cacheKey, expire.toLong(), encoded).await()
            return true
        } catch (e: Exception) {
            application.log.error("Manual cache set with lock failed: $e")
            return false
        } finally {
            if (lockAcquired) {
                try {
                    val luaScript =
                        "if redis.call('get', KEYS[1]) == ARGV[1] then return redis.call('del', KEYS[1]) else return 0 end"
                    redis.eval<Long>(luaScript, ScriptOutputType.INTEGER, arrayOf(lockKey), lockValue).await()
                } catch (e: Exception) {
                    application.log.warn("Failed to release lock in cacheSet: $e")
                }
            }
        }
    }
}

private val CacheServiceAttributeKey = AttributeKey<CacheService>("CacheService")
val Application.cacheService: CacheService
    get() = attributes[CacheServiceAttributeKey]

fun Application.registerCacheService() {
    this.attributes[CacheServiceAttributeKey] = CacheService(this)
}

/**
 * 在 Ktor 的 Application 或 Routing 中直接使用，语法类似 Python 的装饰器
 */
suspend fun <T> Application.cache(
    args: List<Any>,
    expire: Int = 300,
    ignoreNull: Boolean = true,
    excludeArgs: List<Int> = emptyList(),
    keyPrefix: String = "",
    lockTimeout: Int = 30,
    serializer: KSerializer<T>,
    block: suspend () -> T
): T {
    return this.cacheService.cacheGetInternal(
        expire = expire,
        ignoreNull = ignoreNull,
        lockTimeout = lockTimeout,
        fallbackFunc = block,
        cacheKey = this.cacheService.generateCacheKey(args = args, excludeArgs = excludeArgs, keyPrefix = keyPrefix),
        serializer = serializer,
        redis = this.redisManager.getAsyncCommands(),
    )
}

suspend fun <T> RoutingContext.cache(
    args: List<Any>,
    expire: Int = 300,
    ignoreNull: Boolean = true,
    excludeArgs: List<Int> = emptyList(),
    keyPrefix: String = "",
    lockTimeout: Int = 30,
    serializer: KSerializer<T>,
    block: suspend () -> T
): T {
    return call.application.cacheService.cacheGetInternal(
        expire = expire,
        ignoreNull = ignoreNull,
        lockTimeout = lockTimeout,
        fallbackFunc = block,
        cacheKey = call.application.cacheService.generateCacheKey(
            args = args,
            excludeArgs = excludeArgs,
            keyPrefix = keyPrefix
        ),
        serializer = serializer,
        redis = call.application.redisManager.getAsyncCommands(),
    )
}